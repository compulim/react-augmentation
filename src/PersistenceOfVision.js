import PropTypes from 'prop-types';
import React     from 'react';

export default class PersistenceOfVision extends React.PureComponent {
  constructor(props, context) {
    super(props, context);

    this.state = {
      children: props.children
    };
  }

  componentWillReceiveProps(nextProps) {
    if (React.Children.count(nextProps.children)) {
      this.setState(() => ({
        children: nextProps.children
      }));
    } else if (nextProps.children !== this.props.children) {
      this.props.onPersist();
    }
  }

  render() {
    const { onPersist, wrappingType, ...otherProps } = this.props;

    return React.createElement(wrappingType, { ...otherProps }, this.state.children );
  }
}

PersistenceOfVision.defaultProps = {
  onPersist   : () => 0,
  wrappingType: 'div'
};

PersistenceOfVision.propTypes = {
  onPersist   : PropTypes.func,
  wrappingType: PropTypes.oneOfType([PropTypes.func, PropTypes.string])
};
