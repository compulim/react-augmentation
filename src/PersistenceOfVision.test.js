import React    from 'react';
import renderer from 'react-test-renderer';

import PersistenceOfVision from './PersistenceOfVision';

it('passthru when not removed', () => {
  const component = renderer.create(
    <PersistenceOfVision className="passed">
      <p>Hello, World!</p>
    </PersistenceOfVision>
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <PersistenceOfVision className="passed-not-persisted">
      <p>Aloha!</p>
    </PersistenceOfVision>
  );

  expect(component.toJSON()).toMatchSnapshot();
});

it('persist and unpersist correctly', () => {
  const component = renderer.create(
    <PersistenceOfVision>
      <p>Hello, World!</p>
    </PersistenceOfVision>
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <PersistenceOfVision />
  );

  expect(component.toJSON()).toMatchSnapshot();
});

it('initial persist correctly', () => {
  const component = renderer.create(
    <PersistenceOfVision />
  );

  expect(component.toJSON()).toMatchSnapshot();
});

it('passthru and persist multiple children correctly', () => {
  const component = renderer.create(
    <PersistenceOfVision>
      <p>How are you?</p>
      <p>I am fine.</p>
    </PersistenceOfVision>
  );

  expect(component.toJSON()).toMatchSnapshot();

  component.update(
    <PersistenceOfVision />
  );

  expect(component.toJSON()).toMatchSnapshot();
});

it('should emit onPersist when unmount', () => {
  let numHandlePersist = 0;
  const component = renderer.create(
    <PersistenceOfVision onPersist={ () => numHandlePersist++ }>
      <p>Hello, World!</p>
    </PersistenceOfVision>
  );

  component.update(
    <PersistenceOfVision onPersist={ () => numHandlePersist++ } />
  );

  expect(numHandlePersist).toBe(1);
});

it('should use a string wrapping type', () => {
  const component = renderer.create(
    <PersistenceOfVision wrappingType="li">
      Hello, World!
    </PersistenceOfVision>
  );

  expect(component.toJSON()).toMatchSnapshot();
});

it('should use a wrapping component', () => {
  const component = renderer.create(
    <PersistenceOfVision wrappingType={ props => <li>{ props.children }</li> }>
      Hello, World!
    </PersistenceOfVision>
  );

  expect(component.toJSON()).toMatchSnapshot();
});