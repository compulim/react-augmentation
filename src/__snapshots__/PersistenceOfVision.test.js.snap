// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial persist correctly 1`] = `<div />`;

exports[`passthru and persist multiple children correctly 1`] = `
<div>
  <p>
    How are you?
  </p>
  <p>
    I am fine.
  </p>
</div>
`;

exports[`passthru and persist multiple children correctly 2`] = `
<div>
  <p>
    How are you?
  </p>
  <p>
    I am fine.
  </p>
</div>
`;

exports[`passthru when not removed 1`] = `
<div
  className="passed"
>
  <p>
    Hello, World!
  </p>
</div>
`;

exports[`passthru when not removed 2`] = `
<div
  className="passed-not-persisted"
>
  <p>
    Aloha!
  </p>
</div>
`;

exports[`persist and unpersist correctly 1`] = `
<div>
  <p>
    Hello, World!
  </p>
</div>
`;

exports[`persist and unpersist correctly 2`] = `
<div>
  <p>
    Hello, World!
  </p>
</div>
`;

exports[`should use a string wrapping type 1`] = `
<li>
  Hello, World!
</li>
`;

exports[`should use a wrapping component 1`] = `
<li>
  Hello, World!
</li>
`;
